lib/arch/avx2/codec.c

```diff
@@ -13,57 +13,82 @@
 #define REPLACE(s,n)	_mm256_and_si256((s), _mm256_set1_epi8(n))
 #define RANGE(s,a,b)	_mm256_andnot_si256(CMPGT((s), (b)), CMPGT((s), (a) - 1))
 
-static inline __m256i
-_mm256_bswap_epi32 (const __m256i in)
-{
-	// _mm256_shuffle_epi8() works on two 128-bit lanes separately:
-	return _mm256_shuffle_epi8(in, _mm256_setr_epi8(
-		 3,  2,  1,  0,
-		 7,  6,  5,  4,
-		11, 10,  9,  8,
-		15, 14, 13, 12,
-		 3,  2,  1,  0,
-		 7,  6,  5,  4,
-		11, 10,  9,  8,
-		15, 14, 13, 12));
-}
-
-static inline __m256i
-enc_reshuffle (__m256i in)
-{
-	// Spread out 32-bit words over both halves of the input register:
-	in = _mm256_permutevar8x32_epi32(in, _mm256_setr_epi32(
-		0, 1, 2, -1,
-		3, 4, 5, -1));
-
-	// Slice into 32-bit chunks and operate on all chunks in parallel.
-	// All processing is done within the 32-bit chunk. First, shuffle:
-	// before: [eeeeeeff|ccdddddd|bbbbcccc|aaaaaabb]
-	// after:  [00000000|aaaaaabb|bbbbcccc|ccdddddd]
-	in = _mm256_shuffle_epi8(in, _mm256_set_epi8(
-		-1, 9, 10, 11,
-		-1, 6,  7,  8,
-		-1, 3,  4,  5,
-		-1, 0,  1,  2,
-		-1, 9, 10, 11,
-		-1, 6,  7,  8,
-		-1, 3,  4,  5,
-		-1, 0,  1,  2));
-
-	// merged  = [0000aaaa|aabbbbbb|bbbbcccc|ccdddddd]
-	const __m256i merged = _mm256_blend_epi16(_mm256_slli_epi32(in, 4), in, 0x55);
-
-	// bd      = [00000000|00bbbbbb|00000000|00dddddd]
-	const __m256i bd = _mm256_and_si256(merged, _mm256_set1_epi32(0x003F003F));
-
-	// ac      = [00aaaaaa|00000000|00cccccc|00000000]
-	const __m256i ac = _mm256_and_si256(_mm256_slli_epi32(merged, 2), _mm256_set1_epi32(0x3F003F00));
-
-	// indices = [00aaaaaa|00bbbbbb|00cccccc|00dddddd]
-	const __m256i indices = _mm256_or_si256(ac, bd);
-
-	// return  = [00dddddd|00cccccc|00bbbbbb|00aaaaaa]
-	return _mm256_bswap_epi32(indices);
+static inline __m256i enc_reshuffle(const __m256i input) {
+	// input, bytes MSB to LSB:
+	// 0 0 0 0 x w v u t s r q p o n m
+	// l k j i h g f e d c b a 0 0 0 0
+
+	// translation from SSE into AVX2 of procedure
+	// https://github.com/WojciechMula/base64simd/blob/master/encode/unpack_bigendian.cpp
```

Comment:
The hyperlink to the origin is totally fine, but it would be nice if we could get a little context about what the algorithm is _doing_. I'd like this library to stand on its own.